language/metrics-output-plugin/src/metrics-output.ts

```diff
@@ -0,0 +1,178 @@
+    this.outputDir = options.outputDir || process.cwd();
+    this.fileName = options.fileName || "metrics";
+    this.rootProperties = options.rootProperties || {};
+    this.stats = options.stats || {};
+    this.features = options.features || {};
+
+    // Initialize with root properties including timestamp
+    this.aggregatedResults = {
+      timestamp: new Date().toISOString(),
+      ...this.rootProperties,
+      content: {},
+    };
+  }
+
+  apply(service: PlayerLanguageService): void {
+    // Hook into the validation end to capture diagnostics and write output
+    service.hooks.onValidateEnd.tap(
+      this.name,
+      (
+        diagnostics: Diagnostic[],
+        { documentContext }: { documentContext: DocumentContext },
+      ): Diagnostic[] => {
+        this.generateFile(diagnostics, documentContext);
+        return diagnostics;
+      },
+    );
+  }
+
+  /**
+   * Evaluates a value, executing it if it's a function
+   */
+  private evaluateValue(
+    value: any,
+    diagnostics: Diagnostic[],
+    documentContext: DocumentContext,
+  ): any {
+    if (typeof value === "function") {
+      try {
+        return value(diagnostics, documentContext);
+      } catch (error) {
+        documentContext.log.error(`Error evaluating value: ${error}`);
+        return { error: `Value evaluation failed: ${error}` };
+      }
+    }
+    return value;
+  }
+
+  private generateMetrics(
+    diagnostics: Diagnostic[],
+    documentContext: DocumentContext,
+  ): Record<string, any> {
+    const result: Record<string, any> = {};
+
+    // Process each metric
+    Object.entries(this.stats).forEach(([key, value]) => {
+      result[key] = this.evaluateValue(value, diagnostics, documentContext);
+    });
+
+    return result;
+  }
+
+  private generateFeatures(
+    diagnostics: Diagnostic[],
+    documentContext: DocumentContext,
+  ): Record<string, any> {
+    const result: Record<string, any> = {};
+
+    // Process each feature
+    Object.entries(this.features).forEach(([key, value]) => {
+      result[key] = this.evaluateValue(value, diagnostics, documentContext);
+    });
+
+    return result;
+  }
+
+  private generateFile(
+    diagnostics: Diagnostic[],
+    documentContext: DocumentContext,
+  ): string {
+    // Ensure the output directory exists
+    const fullOutputDir = path.resolve(process.cwd(), this.outputDir);
+    fs.mkdirSync(fullOutputDir, { recursive: true });
+
+    // Get the file path from the document URI
+    const filePath = documentContext.document.uri;
+
+    // Generate metrics
+    const stats = this.generateMetrics(diagnostics, documentContext);
+    const features = this.generateFeatures(diagnostics, documentContext);
+
+    this.aggregatedResults.content[filePath] = {
+      stats,
+      features: Object.keys(features).length > 0 ? features : {},
+    };
+
+    // Write the aggregated results to a file
+    const outputFilePath = path.join(fullOutputDir, `${this.fileName}.json`);
```

Comment:
if we're going to add the `.json` automatically we should document that just so people don't pass `filename.json` as the input

Reply:
You raise a good point, I wonder if we should just have it be the full file name

Reply:
In addition to docs, also updated the JSDoc